arrow/src/compute/kernels/interleave.rs

```
@@ -0,0 +1,218 @@
+use arrow_schema::ArrowError;
+
+///
+/// Takes elements by index from a list of [`Array`], creating a new [`Array`] from those values.
+///
+/// Each element in `indices` is a pair of `usize` with the first identifying the index
+/// of the [`Array`] in `values`, and the second the index of the value within that [`Array`]
+///
+/// ```text
+/// ┌─────────────────┐      ┌─────────┐                                  ┌─────────────────┐
+/// │        A        │      │ (0, 0)  │        interleave(               │        A        │
+/// ├─────────────────┤      ├─────────┤          [values0, values1],     ├─────────────────┤
+/// │        D        │      │ (1, 0)  │          indices                 │        B        │
+/// └─────────────────┘      ├─────────┤        )                         ├─────────────────┤
+///   values array 0         │ (1, 1)  │      ─────────────────────────▶  │        C        │
+///                          ├─────────┤                                  ├─────────────────┤
+///                          │ (0, 1)  │                                  │        D        │
+///                          └─────────┘                                  └─────────────────┘
+/// ┌─────────────────┐       indices
+/// │        B        │        array
+/// ├─────────────────┤                                                    result
+/// │        C        │
+/// ├─────────────────┤
+/// │        E        │
+/// └─────────────────┘
+///   values array 1
+/// ```
+///
+/// For selecting values by index from a single array see [compute::take](crate::compute::take)
+///
+/// # Panics
+///
+/// Panics if the arrays do not have the same data type or `values` is empty
+pub fn interleave(
+    values: &[&dyn Array],
+    indices: &[(usize, usize)],
+) -> Result<ArrayRef, ArrowError> {
+    if values.is_empty() {
+        return Err(ArrowError::InvalidArgumentError(
+            "interleave requires input of at least one array".to_string(),
+        ));
+    }
```

Should we also return an error for single array case and suggest to use `compute::take`?

---

I vacillated a bit on this, I think given the concat kernel which makes even less sense to be called on a single array, doesn't error in this case - I would be inclined to leave it.

---

I agree that it is nicer to support handling a single input (mostly as a convenience for downstream users so they don't have to special case their code for `len() = 1` case)